### ISSUE TITLE
DataObject is being written twice when saving from GridFieldDetailForm

### ISSUE THREAD

**Issue by ElGabbu** [silverstripe/silverstripe-framework#5879]
When adding/editing a dataObject of a HasMany relationship, during saving of the form the dataObject is being written twice, and thus calling the onAfterWrite twice. 

The reason is because in the GridFieldDetailForm_ItemRequest 

```
$this->record->write();
$list->add($this->record, $extraData);
```

Since HasManyList::add calls write again on the record/dataobject we end up with two invokes of onAfterWrite. 

A possible solution would be:

```
if (!$list instanceof HasManyList) { 
   $this->record->write(); 
}
$list->add($this->record, $extraData);
```

Thanks

**Comment by dhensby** [comment]
The second write shouldn't touch the DB because there will be no changes to write...

I don't think this is a bug as it's somewhat intended behaviour. The `GridFieldDetailForm_ItemRequest` shouldn't be making assumptions about whether the `->add` method performs a write or not. It needs to take care of it's responsibilities (which is writing the object).

**Comment by kinglozzer** [comment]
@dhensby I believe the second change will be the `HasOneID` on `$this->record` which changes when `HasManyList::add()` is called. That probably explains why `onAfterWrite()` is still being called.

I don’t like that we’re triggering unnecessary writes. I know that one write in this context is completely insignificant from a performance point of view, but I still don’t like it. I just can’t think of a cleaner way to “fix” this 😞

**Comment by ElGabbu** [comment]
@kinglozzer ... on AfterWrite is still called even if the hasOneID doesn't change. 

@dhensby yes true but the second write will invoke onAfterWrite hooks nonetheless ... which is problematic for the logic I want to put in my onAfterWrite hook.

**Comment by kinglozzer** [comment]
@ElGabbu Which version of SilverStripe are you using? In 3.2+, it should only trigger `onAfterWrite` if something has actually changed (https://github.com/silverstripe/silverstripe-framework/blob/3.2/model/DataObject.php#L1382-L1400). I’m not sure about 3.1.

**Comment by ElGabbu** [comment]
@kinglozzer yeah this project is on 3.1.18 .... good to know it was fixed in 3.2

**Comment by dhensby** [comment]
It's a shame that a second write does hit the DB if the relation ID changes, but ultimately we definitely shouldn't be making assumptions about who has responsibility for the writes. If it means more than one write, then so be it.

If we see two `onBeforeWrites` that's because we're seeing changes twice, I get that this is a pain, but one should be careful what they use this hook for and not rely on it to only ever be called once or making assumptions about what has changed.
